src/enlyze/client.py

```diff
@@ -150,6 +187,95 @@ def get_variables(
             for variable in self._get_variables(machine.uuid)
         ]
 
+    def _get_paginated_timeseries(
+        self,
+        *,
+        machine_uuid: str,
+        start: datetime,
+        end: datetime,
+        variables: Sequence[str],
+        resampling_interval: Optional[int],
+    ) -> Iterator[timeseries_api_models.TimeseriesData]:
+        params: dict[str, Any] = {
+            "appliance": machine_uuid,
+            "start_datetime": start.isoformat(),
+            "end_datetime": end.isoformat(),
+            "variables": ",".join(variables),
+        }
+
+        if resampling_interval:
+            params["resampling_interval"] = resampling_interval
+
+        return self._timeseries_api_client.get_paginated(
+            "timeseries", timeseries_api_models.TimeseriesData, params=params
+        )
+
+    def _get_timeseries(
+        self,
+        start: datetime,
+        end: datetime,
+        variables: Union[
+            Sequence[user_models.Variable],
+            Mapping[user_models.Variable, user_models.ResamplingMethod],
+        ],
+        resampling_interval: Optional[int] = None,
+    ) -> Optional[user_models.TimeseriesData]:
+        variables_sequence, variables_query_parameter_list = (
+            _get_variables_sequence_and_query_parameter_list(
+                variables, resampling_interval
+            )
+        )
+
+        start, end, machine_uuid = validate_timeseries_arguments(
+            start, end, variables_sequence
+        )
+
+        try:
+            chunks = chunk(
+                variables_query_parameter_list,
+                MAXIMUM_NUMBER_OF_VARIABLES_PER_TIMESERIES_REQUEST,
+            )
+        except ValueError as e:
+            raise EnlyzeError(FETCHING_TIMESERIES_DATA_ERROR_MSG) from e
+
+        chunks_pages = (
+            self._get_paginated_timeseries(
+                machine_uuid=machine_uuid,
+                start=start,
+                end=end,
+                variables=chunk,
+                resampling_interval=resampling_interval,
+            )
+            for chunk in chunks
+        )
+
+        timeseries_data_chunked = [
+            _get_timeseries_data_from_pages(pages) for pages in chunks_pages
+        ]
+
+        if not timeseries_data_chunked or all(
+            data is None for data in timeseries_data_chunked
+        ):
+            return None
+
+        if any(data is None for data in timeseries_data_chunked) and any(
+            data is not None for data in timeseries_data_chunked
+        ):
+            raise EnlyzeError(
+                "The timeseries API didn't return data for some of the variables."
+            )
+
+        try:
+            timeseries_data = reduce(lambda x, y: x.merge(y), timeseries_data_chunked)  # type: ignore # noqa
+        except ValueError as e:
+            raise EnlyzeError(FETCHING_TIMESERIES_DATA_ERROR_MSG) from e
```

Comment:
re [your comment](https://github.com/enlyze/enlyze-python/pull/40/files#r1502337215):

```suggestion
        try:
            first_chunk_pages = next(chunks_pages)
        except StopIteration:
            return None

        timeseries_data = _get_timeseries_data_from_pages(first_chunk_pages)
        if timeseries_data is None:
            return None

        for chunk_pages in chunks_pages:
            chunk_data = _get_timeseries_data_from_pages(chunk_pages)
            if not chunk_data:
                raise EnlyzeError(
                    "The timeseries API didn't return data for some of the variables."
                )

            try:
                timeseries_data.merge(chunk_data)
            except ValueError as e:
                raise EnlyzeError(FETCHING_TIMESERIES_DATA_ERROR_MSG) from e
```

Reply:
This yields little to no improvement in terms of memory usage over the other code. The difference here is that you're not maintaining a reference to the object _(albeit the same amount of objects is still getting created)_ created in 

```python
chunk_data = _get_timeseries_data_from_pages(chunk_pages)
```

But this doesn't mean that this memory is freed by the interpreter because most likely the GC won't free this memory until its next pass (most likely won't be before the scope of the function ends). You can verify that if you'd like by running both code snippets under a memory profiler.

> There's no limit imposed by the SDK w.r.t. the amount of data we're pulling from the API, so it can easily be hundreds of megabytes when fetching longer time frames. 
> 
> FYI: we already use the SDK in our own services, so it has a direct impact on our cluster as well.

AFAIK, I am the only one who encountered the issue that resulted in this PR while I was doing some tests. If that's an issue that we run into regularly we'll be able to find that out by memory usage alerts in the services, but as far as I am concerned, all this is premature optimization right now and if we'd want to tackle it we'd need to change the code in a non-trivial manner. We can also discuss introducing limits on the amount of data that the SDK can fetch at once.

Reply:
> This yields little to no improvement in terms of memory usage over the other code. The difference here is that you're not maintaining a reference to the object _(albeit the same amount of objects is still getting created)_ created in
> 
> ```python
> chunk_data = _get_timeseries_data_from_pages(chunk_pages)
> ```
> 
> But this doesn't mean that this memory is freed by the interpreter because most likely the GC won't free this memory until its next pass (most likely won't be before the scope of the function ends). You can verify that if you'd like by running both code snippets under a memory profiler.

Thanks for the explanation. Indeed my (intuitive) thinking was lead by the C/C++ memory model of stack variables (when it leaves scope, memory is freed). I haven't checked with a memory profiler, but you will likely be right that the memory usage will be the same in almost all cases.